tests/templates/fetching_ddp_dynbatch_finetuning/finetune.py

```diff
@@ -0,0 +1,349 @@
+
+    # DDP init
+    sb.utils.distributed.ddp_init_group(run_opts)
+
+    with open(hparams_file) as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
+    # Kept aside for later
+    reporting = {
+        "WER": {
+            "field": "error_rate",
+            "tracker": deepcopy(
+                hparams["error_rate_computer"]()
+            ),  # n_jobs=int(os.environ['WORLD_SIZE']))),
+        },
+        "CER": {
+            "field": "error_rate",
+            "tracker": deepcopy(
+                hparams["cer_computer"]()
+            ),  # n_jobs=int(os.environ['WORLD_SIZE']))),
+        },
+    }
+
+    # Create experiment directory
+    sb.create_experiment_directory(
+        experiment_directory=hparams["output_folder"],
+        hyperparams_to_save=hparams_file,
+        overrides=overrides,
+    )
+
+    # Spared: dataset prepare script & its run_on_main - prepare dataio pipeline only
+    datasets = dataio_prepare(hparams)
+
+    # Add: dynamic batching (only in training & validation)
+    dynamic_hparams = hparams["dynamic_batch_sampler"]
+    train_dataloader_opts = hparams["train_dataloader_opts"]
+    valid_dataloader_opts = hparams["valid_dataloader_opts"]
+    train_dataloader_opts = {
+        "batch_sampler": DynamicBatchSampler(
+            datasets["train"],
+            dynamic_hparams["max_batch_len"],
+            num_buckets=dynamic_hparams["num_buckets"],
+            length_func=lambda x: x["length"],
+            shuffle=dynamic_hparams["shuffle_ex"],
+            batch_ordering=dynamic_hparams["batch_ordering"],
+        ),
+        "num_workers": train_dataloader_opts["num_workers"],
+    }
+    valid_dataloader_opts = {
+        "batch_sampler": DynamicBatchSampler(
+            datasets["valid"],
+            dynamic_hparams["max_batch_len"],
+            num_buckets=dynamic_hparams["num_buckets"],
+            length_func=lambda x: x["length"],
+            shuffle=dynamic_hparams["shuffle_ex"],
+            batch_ordering=dynamic_hparams["batch_ordering"],
+        ),
+        "num_workers": valid_dataloader_opts["num_workers"],
+    }
+
+    # We download:
+    # * the tokenizer from URL - https://huggingface.co/speechbrain/asr-crdnn-rnnlm-librispeech/resolve/main/...
+    # * the pretrained LM from HuggingFace - HF: speechbrain/asr-crdnn-rnnlm-librispeech
+    # * the pretrained ASR from the local template checkpoint - local: speechbrain/asr-crdnn-rnnlm-librispeech
+    run_on_main(
+        hparams["pretrainer_tokenizer"].collect_files,
+        kwargs={"fetch_from": FetchFrom.ONLINE},
+    )
+    run_on_main(
+        hparams["pretrainer_LM"].collect_files,
+        kwargs={"fetch_from": FetchFrom.HUGGING_FACE},
+    )
+    hparams["pretrainer_tokenizer"].load_collected(run_opts["device"])
+    hparams["pretrainer_LM"].load_collected(run_opts["device"])
+    # LOCAL fetching takes sources directly from their location
+    hparams["pretrainer_ASR"].collect_files(fetch_from=FetchFrom.LOCAL)
```

Comment:
Here, the same code cannot handle both local and online sources in distributed setup. Local must be called without `run_on_main`, online must be called with it.

Reply:
The `run_on_main` wrapping might get easily overused and applied stacked as one wrapper on another one.

I'd agree with moving such wrappers to the instance when they are needed, and not as a general thing to do for each recipe. It must look somewhat absurd to users at some point to always have to put this. Or they just do and move on. Implementing that change is a major refactoring. cc @TParcollet @mravanelli 

@Gastron is this comment good/bad or an observation? ^^"

Reply:
I think there are some conflicting ideals in this:

1. Like you mentioned, adding this wrapper is an additional burden on the user.
2. However, something getting run on the main and not on other processes can be surprising,
3. and library code should ideally deal with a single responsibility. Adding multi-process orchestration to that is another extra responsibility IMO.

Perhaps we could have come up with some name suffix (like `collect_files_ddp_compliant`) that would signal a version of a method that takes care of multi-process issues - that would deal with the surprise by signaling in the name, and this special version of the method would take of the orchestration, using the non-wrapped code as the implementation, so it would again be single-responsibility-principle-good.

Reply:
But yeah changing it everywhere would be a bigger task. But I am also OK with not going through with a major refactor, perhaps we can just put `run_on_main` into collect_files in this one case and leave it there for now.